com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaHotCodeReplaceProvider.java

```diff
@@ -0,0 +1,343 @@
+ * Copyright (c) 2017 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Microsoft Corporation - Adapter the code for VSCode Java Debugger
+ *******************************************************************************/
+
+package com.microsoft.java.debug.plugin.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.IResourceDeltaVisitor;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaModelMarker;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.ToolFactory;
+import org.eclipse.jdt.core.util.IClassFileReader;
+import org.eclipse.jdt.core.util.ISourceAttribute;
+
+import com.microsoft.java.debug.core.Configuration;
+import com.microsoft.java.debug.core.adapter.IDebugAdapterContext;
+import com.microsoft.java.debug.core.adapter.IHotCodeReplaceProvider;
+import com.microsoft.java.debug.core.protocol.Events.DebugEvent;
+
+/**
+ * The hot code replace provider listens for changes to class files and notifies
+ * the running debug session of the changes.
+ */
+public class JavaHotCodeReplaceProvider implements IHotCodeReplaceProvider, IResourceChangeListener {
+
+    private static final Logger logger = Logger.getLogger(Configuration.LOGGER_NAME);
+
+    private static final String CLASS_FILE_EXTENSION = "class"; //$NON-NLS-1$
+
+    private AtomicBoolean needHotCodeReplace = new AtomicBoolean(false);
+
+    /**
+     * Visitor for resource deltas.
+     */
+    protected ChangedClassFilesVisitor classFilesVisitor = new ChangedClassFilesVisitor();
+
+    /**
+     * A visitor which collects changed class files.
+     */
+    class ChangedClassFilesVisitor implements IResourceDeltaVisitor {
+        /**
+         * The collection of changed class files.
+         */
+        protected List<IResource> changedFiles = null;
+
+        /**
+         * Collection of qualified type names, corresponding to class files.
+         */
+        protected List<String> fullyQualifiedNames = null;
```

Comment:
Where are initialize changedFile and fullyQualifiedNames initialized?

Reply:
TBA in another change